src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +122,153 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
+            }
+        }
+
+        internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, DiagnosticBag diagnosticsBag, CancellationToken cancellationToken = default)
+        {
+            // with no generators, there is no work to do
+            if (_state.Generators.IsEmpty)
+            {
+                return _state;
+            }
+
+            // run the actual generation
+            var state = StateWithPendingEditsApplied(_state);
+            var stateBuilder = ArrayBuilder<GeneratorState>.GetInstance();
+            int receiverCount = 0;
+
+            for (int i = 0; i < state.Generators.Length; i++)
+            {
+                var generator = state.Generators[i];
+                var generatorState = state.GeneratorStates[i];
+
+                // initialize the generator if needed
+                if (!generatorState.Info.Initialized)
+                {
+                    InitializationContext context = new InitializationContext(cancellationToken);
+                    Exception? ex = null;
+                    try
+                    {
+                        generator.Initialize(context);
+                    }
+                    catch (Exception e)
+                    {
+                        ex = e;
+                    }
+                    generatorState = ex is null
+                                     ? new GeneratorState(context.InfoBuilder.ToImmutable())
+                                     : SetGeneratorException(new GeneratorState(), generator, ex, diagnosticsBag, isInit: true);
+                }
+
+                // create the syntax receiver if requested
+                if (generatorState.Info.SyntaxReceiverCreator is object)
+                {
+                    ISyntaxReceiver? rx = null;
+                    try
+                    {
+                        rx = generatorState.Info.SyntaxReceiverCreator();
+                    }
+                    catch (Exception e)
+                    {
+                        generatorState = SetGeneratorException(generatorState, generator, e, diagnosticsBag);
+                    }
+
+                    if (rx is object)
+                    {
+                        generatorState = generatorState.WithReceiver(rx);
+                        receiverCount++;
+                    }
+                }
+
+                stateBuilder.Add(generatorState);
+            }
+
+            // Run a syntax walk if any of the generators requested it
+            if (receiverCount > 0)
+            {
+                for (int i = 0; i < stateBuilder.Count; i++)
+                {
+                    var generatorState = stateBuilder[i];
+                    if (generatorState.SyntaxReceiver is object)
+                    {
+                        GeneratorSyntaxWalker walker = new GeneratorSyntaxWalker(generatorState.SyntaxReceiver);
+                        foreach (var tree in compilation.SyntaxTrees)
```

Comment:
Is this going to do an entire walk of the compilation for every syntax receiver? I think I would likely expect there to be more syntax trees than generators, so it seems like inverting this ordering should be better.

Reply:
If we inverted the ordering, I think it would be possible to parallelize the running of the receivers on every tree as well.

---
In reply to: [466073588](https://github.com/dotnet/roslyn/pull/46296#discussion_r466073588) [](ancestors = 466073588)

Reply:
So, we used to do it the other way around, and I actually implemented it working that way, but the code was super confusing. Then thinking it through I couldn't convince myself that calling each generator per syntax node was actually any more efficient that doing a walk per generator. Ultimately there are O(n x m) operations? intuitively that felt wrong but I wasn't able to convince myself.

---
In reply to: [466074746](https://github.com/dotnet/roslyn/pull/46296#discussion_r466074746) [](ancestors = 466074746,466073588)

Reply:
I think that inverting the iteration order makes more sense to me. You can more easily parallelize that implementation later: I don't think we can ask writers to write their receivers to expect visiting multiple nodes at the same time, and as the innermost loop that's what we would be easiest to parallelize with this implementation. However, with an inverted structure, we can easily run all of the receivers over the same tree at the same time. We could also have the creation of the `stateBuilder` create the walkers for each receiver, rather than creating it here, which would make object reuse better. To list the steps:
1. Loop over the entire list of generators. Initialize if necessary, and create the tree walker for the receiver, if present.
2. Loop over the trees in the compilation.
    a. Loop over all generators, and have them visit the current tree. This is the part that is easily parallelizable.
3. Remaining work as today.

---
In reply to: [466083669](https://github.com/dotnet/roslyn/pull/46296#discussion_r466083669) [](ancestors = 466083669,466074746,466073588)

Reply:
Ah, I see what you're saying and that makes sense. It used to be even more flipped, where each syntax node visit called each generator, so we only did a single walk of each tree, but thats complicated, and yeah means receivers getting trees in odd orders. I think your suggestion above is a happy medium between the two. 

(Note on parallelization: this is an interesting topic, for the compiler its trivial and obvious to do it in parallel. for the IDE it can make perf worse because you can't reason about thread usage).

---
In reply to: [466572237](https://github.com/dotnet/roslyn/pull/46296#discussion_r466572237) [](ancestors = 466572237,466083669,466074746,466073588)